src/macos_native.rs

```diff
@@ -0,0 +1,892 @@
+    let mut input_report_buffer = vec![0u8; shared_state.max_input_report_len];
+
+    // This must live as long as the callback is registered
+    let ctx_ptr = Arc::as_ptr(&shared_state) as *const c_void as *mut c_void;
+
+    let input_report_context = shared_state.clone();
+
+    let run_loop_mode = CFString::new(&shared_state.run_loop_mode);
+
+    let mut _input_report_callback = None;
+
+    {
+        let device = shared_state.device.lock().unwrap();
+
+        _input_report_callback = Some(device.register_input_report_callback(
+            &mut input_report_buffer,
+            Some(hid_report_callback),
+            input_report_context,
+        ));
+
+        unsafe { device.register_removal_callback(Some(hid_removal_callback), ctx_ptr) }
+
+        device.schedule_with_run_loop(&CFRunLoop::get_current(), &run_loop_mode);
+
+        let mut ctx = CFRunLoopSourceContext {
+            version: 0,
+            info: ctx_ptr as *const c_void as *mut _,
+            retain: None,
+            release: None,
+            copyDescription: None,
+            equal: None,
+            hash: None,
+            schedule: None,
+            cancel: None,
+            perform: perform_signal_callback,
+        };
+
+        let source = unsafe {
+            CFRunLoopSourceCreate(kCFAllocatorDefault, 0 /* order */, &mut ctx)
+        };
+
+        let source = unsafe { CFRunLoopSource::wrap_under_create_rule(source) };
+
+        let current_run_loop = CFRunLoop::get_current();
+
+        current_run_loop.add_source(&source, run_loop_mode.as_concrete_TypeRef());
+
+        let mut shared_source = shared_state.source.lock().unwrap();
+        *shared_source = Some(LoopSource(source));
+    }
+
+    {
+        let mut run_loop = shared_state.run_loop.lock().unwrap();
+
+        *run_loop = Some(WrappedCFRunLoop(CFRunLoop::get_current()));
+    }
+
+    barrier.wait();
+
+    while (!shared_state.shutdown_thread.load(Ordering::Relaxed))
+        && (!shared_state.disconnected.load(Ordering::Relaxed))
+    {
+        // TODO: Verify timeout value
+        let code = unsafe { CFRunLoopRunInMode(run_loop_mode.as_concrete_TypeRef(), 1000.0, 0) };
+
+        // Return if the device has been disconnected
+        if code == CFRunLoopRunResult::Finished as i32 {
+            shared_state.disconnected.store(true, Ordering::Relaxed);
+            break;
+        }
+
+        // Break if the run loop returns finished or stopped
+        if code != CFRunLoopRunResult::HandledSource as i32
+            && code != CFRunLoopRunResult::TimedOut as i32
+        {
+            shared_state.shutdown_thread.store(true, Ordering::Relaxed);
+            break;
+        }
+    }
+
+    // Notify that the thread is stopping
+    {
+        let _guard = shared_state.input_reports.lock().unwrap();
+        shared_state.condition.notify_all();
+    }
+
+    {
+        let device = shared_state.device.lock().unwrap();
+
+        // unregister the input report callback
```

Comment:
This isn't necessary - `IOHIDDevice` automatically unregisters these on close: https://github.com/apple-oss-distributions/IOKitUser/blob/b0b3f822b7507c265aa8a1e37c3100c03ca82039/hid.subproj/IOHIDDevice.c#L483-L486